[BEAM-135] Utilities for "batching" elements in a DoFn

echauchot opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).
R: @kennknowles 
---
**This work is currently in progress, the tests don't pass yet. This PR was created to have a convenient environment for discussing this code. Here are the TODOS:**

- windows after applying the transform do not contain the right number of elements. They only contain the elements that were output when batchSize was reached and not the ones that are left in the window if batchSize < windowSize. These remaining elements are supposed to be output when the window ends thanks to a timer that fires at window.maxTimestamp + allowedLateness. Elements are added to a bagState but this bagState turns out to be empty when the timer fires. Maybe, as the states are scoped per window, and the window is different (end of fixedWindow passing to GlobalWindow), then another bagState instance gets injected. Hence the empty bagState? 

- Test of the inner BatchingDoFn is ignored because DoFnTester is not supported by DirectRunner yet.

dhalperi commented:
R: @kennknowles @tgroh 

Kenn (state & timers) an Thomas (direct runner) can likely help debug here.

kennknowles commented:
Just to keep this up-to-date, there are definitely multiple bugs in how I implemented timers in the direct runner. @tgroh is doing it correctly now.

kennknowles commented:
The known issues with timers are now resolved at HEAD.

echauchot commented:
Cool! Thank you guys.

echauchot commented:
@kennknowles, I just rebased on master to profit from #2285 and it now works.
I'm gonna clean the debug logs, and I'll ping you for a review of this PR

echauchot commented:
I did some cleaning, and squashed into fewer commits. Ready for review.
R: @kennknowles

echauchot commented:
R: @bjchambers, I forgot to add you as a reviewer

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+                    batchSize,
+                    perBatchFn,
+                    allowedLateness,
+                    (Coder<InputT>) input.getCoder().getCoderArguments().get(1),
```
I would prefer to `checkState` that the input coder is a `KvCoder`, cast to it, and explicitly extract the `keyCoder` and `valueCoder` from it. Any failure would be much simpler to debug.

kennknowles replied:
`checkArgument` :-)

echauchot replied:
+1

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+    private final TimerSpec timer = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+    @StateId(TIMER_ALREADY_SET_ID)
+    private final StateSpec<Object, ValueState<Integer>> timerAlreadySetForWindow;
```
You shouldn't need this. Blind-writing a timer should be correct even if that timer is already set, and faster than checking if the timer has already been set.

echauchot replied:
just a typo, thanks

echauchot replied:
yes, sure it is a leftover of the time I use setForNow instead of set on the timer. I Forgot to remove it

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+    private final StateSpec<Object, ValueState<Long>> numElementsInBatchSpec;
+
+    @StateId(KEY_ID)
+    private final StateSpec<Object, ValueState<K>> keySpec;
```
It's really unfortunate that you have to do this (but I believe currently required). I've filed https://issues.apache.org/jira/browse/BEAM-1819 to stop you from having to, but until that's resolved you'll have to continue doing this

echauchot replied:
Yes, I was surprised having to do this :) Thanks for the jira ticket!
 Keep me posted when the PR is done and merged.

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+    private final StateSpec<Object, BagState<InputT>> batchSpec;
+
+    @StateId(NUM_ELEMENTS_IN_BATCH_ID)
+    private final StateSpec<Object, ValueState<Long>> numElementsInBatchSpec;
```
This can be a `combiningValue` state, which I believe more accurately models the way in which you actually use the state - you add to the state, and then read-back to see if you're done with a batch.

kennknowles replied:
This is also more efficient, as some runners will blind write a bag of accumulators and only combine them on final read.

echauchot replied:
Yes, thanks for the suggestion !

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+      }
+      batch.clear();
+      LOGGER.debug("*** BATCH *** clear");
+      numElementsInBatch.write(0L);
```
Ideally this would also be `clear()`. If you use `CombiningValueState`, I believe it can be.

kennknowles replied:
It can be `clear()` either way.

echauchot replied:
ok

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,284 @@
+    return data;
+  }
+
+  // timer is not supported by DoFnTester
```
Should be the value to `@Ignore`

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,284 @@
+  private static final int TIMESTAMP_INTERVAL = 1;
+  private static final long WINDOW_DURATION = 5;
+  private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(BatchingParDoTest.class);
+  private static ArrayList<TimestampedValue<KV<String, String>>> data;
```
These probably don't need to be static. Additionally, you should be able to inline their initialization with the variable declaration.

echauchot replied:
Missed that one ! Indeed I moved the initialization from the @BeforeClass to the variable declaration to avoid the need for static and I put it transient to avoid serialization problems.

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,284 @@
+  }
+
+  @Test
+  @Category({RunnableOnService.class, UsesTimersInParDo.class})
```
Also `UsesStatefulParDo.class`, yes?

echauchot replied:
I have seen this category and I asked myself the same question. 
But now that I look at the pom of spark runner (that does not support state yet), for the runnable-on-service-tests, I see UsesStatefulParDo category in the exclusions. 
I added the category in BatchingParDo

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,284 @@
+
+  @Test
+  @Category({RunnableOnService.class, UsesTimersInParDo.class})
+  public void testInBatchMode() {
```
This is more `testGloballyWindowed`, yes?

echauchot replied:
yes better name indeed

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,284 @@
+            .apply("Input data", Create.of(data))
+            // remove timestamps from dataset to be closer to users usecase
+            .apply(
+                "remove timestamps",
```
You can drop this if you `Create.timestamped` above

echauchot replied:
Sure. done

tgroh commented on sdks/java/core/src/test/resources/log4j.properties:
```diff
@@ -0,0 +1,28 @@
+#
```
I don't know if this fits in this PR. It may be valuable, but should have broader discussion.

echauchot replied:
Totally agree, just forgot to remove it

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,284 @@
+    }
+    TestStream<KV<String, String>> stream =
+        streamBuilder
+            .advanceWatermarkTo(startInstant.plus(Duration.standardSeconds(WINDOW_DURATION - 1)))
```
Can `WINDOW_DURATION` be inlined into this test? I don't see it used elsewhere.

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,284 @@
+  @Test
+  @Category({RunnableOnService.class, UsesTimersInParDo.class, UsesTestStream.class})
+  public void testInStreamingMode() {
+    TestStream.Builder<KV<String, String>> streamBuilder =
```
as an aside, this pleases me.

echauchot replied:
Very useful class :1st_place_medal:

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,284 @@
+    }
+  }
+
+  private class CheckValuesFn implements SerializableFunction<Iterable<KV<String, Long>>, Void> {
```
Can this be inlined into the appropriate test? You use it twice in that test, with the same configuration both times.

echauchot replied:
Sure

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,284 @@
+    pipeline.run().waitUntilFinish();
+  }
+
+  private boolean checkAllElementsProcessing(
```
I don't understand what this is actually checking. A comment would be welcome, at minimum.

echauchot replied:
ok, will add javadoc

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,284 @@
+            return output;
+          }
+        };
+    batchingParDo = BatchingParDo.via(BATCH_SIZE, perBatchFn);
```
I would inline the creation of this ParDo; I find it makes tests easier to read, and it's quite low overhead.

echauchot replied:
You mean that the gain of readability is worth creating ParDo twice? Well in a test, I guess creating twice the ParDo is ok

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+      batch.add(c.element().getValue());
+      LOGGER.debug("*** BATCH *** Add element for window %s ", window.toString());
+      Long num = numElementsInBatch.read();
+      if (num == null) {
```
Could be `long num = firstNonNull(numElementsInBatch.read(), 0L)`

echauchot replied:
No problem Kenn, thanks for having taken time for this review !

echauchot commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+  @VisibleForTesting
+  static class BatchingDoFn<K, InputT, OutputT> extends DoFn<KV<K, InputT>, KV<K, OutputT>> {
+
+    private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(BatchingDoFn.class);
```
just a typo, thanks

echauchot replied:
just a typo, thanks

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+ * {@link PTransform} that allows to compute elements in batch of desired size. The input {@link
+ * PCollection} needs to be a {@code PCollection<KV>}. Elements that must belong to the same batch
+ * need to have the same key. Elements are added to a buffer. When the buffer reaches {@code
+ * batchSize}, it is then processed through a user {@link SimpleFunction perBatchFn} function. The
```
It seems like it would be more composable to just output the batch. After that a user can add a SimpleFunction, or an arbitrary DoFn to process the batch.

echauchot replied:
I agree, letting the user deal with the batch with a `PTransform` or `DoFn` seems to offer him more possibilities than using a `SimpleFunction` . I will make the changes.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+ * batchSize}, it is then processed through a user {@link SimpleFunction perBatchFn} function. The
+ * output elements then are added to the output {@link PCollection}. Windows are preserved (batches
+ * contain elements from the same window). Batching is done trans-bundles (batches may contain
+ * elements from more than one bundle)
```
Drop the "Batching is done trans-bundles" and just say "Batches may contain elements from more than one bundle."

echauchot replied:
ok

echauchot commented:
Thank you guys for the review

echauchot commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+public class BatchingParDo<K, InputT, OutputT>
+    extends PTransform<PCollection<KV<K, InputT>>, PCollection<KV<K, OutputT>>> {
+
+  private final long batchSize;
```
I agree with the point on latency. 

Nevertheless the orientation of this feature was to provide exact batching (https://docs.google.com/document/d/1zf9TxIOsZf_fz86TGaiAQqdNI5OO7Sc6qFsxZlBAMiA/edit?usp=sharing) to allow use cases such as abuse detection in an auction system: raise an alert when a user has made exactly 10 auctions in the last hour. The state API allows to count to 10 even if a bundle is smaller than 10 elements. There are still corner cases in which, the exact batching cannot be obtained: if the window contains 15 elements, then there will be a batch of 10 elements and a batch of 5 elements because we need to output the batched elements in the same window as the input.

Regarding what you propose: if we flush batch at the end of the bundle, no need to use state then. And for the use of minBatchSize, it will be little often guaranteed: if after a batch flush, there is fewer than minBatchSize elements left in the window, then, to maintain windowing, we will output less than minBatchSize elements. That is why you said "targeted" I guess. Is this ok for a user to have min and max parameters that are not almost always guaranteed, or at least often?

Maybe it is at the end a compromise to find between possible use cases and latency?
WDYT?

tgroh replied:
I'm forgetting here that it's not possible to flush at the end of the bundle and maintain proper windowing. You're right, ignore me.

echauchot commented:
@tgroh, @kennknowles, @bjchambers, thank you guys for the review! I have addressed all your comments.

There are big changes following Ben's comment (see above), the output is now the batch itself, so I re-coded all the tests. Notably, testUnderlyingDoFn is not needed anymore because now that we output batches, we can test the batching process from outside the ParDo.

 Can you take another look?

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link PTransform} that allows to compute elements in batch of desired size. The input {@link
```
A {@link PTransform} that outputs elements in batches of a desired size. ...

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+
+/**
+ * {@link PTransform} that allows to compute elements in batch of desired size. The input {@link
+ * PCollection} needs to be a {@code PCollection<KV>}. Elements that must belong to the same batch
```
Batches will contain only elements of a single key.

tgroh replied:
I don't think you need the "The input needs to be...", as it's restricted by your type signature.

"Elements within a batch will share a key."

echauchot replied:
I agree. I prefer your sentence "Batches will contain only elements of a single key", will use this one

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,214 @@
+/**
+ * {@link PTransform} that allows to compute elements in batch of desired size. The input {@link
+ * PCollection} needs to be a {@code PCollection<KV>}. Elements that must belong to the same batch
+ * need to have the same key. Elements are added to a buffer. When the buffer reaches {@code
```
I don't think you need to note "Elements are added to a buffer". Perhaps "Elements are buffered until there are {@code batchSize} elements buffered, at which point they are output to the output {@link PCollection"}.

echauchot replied:
Yes, it's a leftover from previous "hidden" buffer architecture

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,225 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link PTransform} that allows to compute elements in batch of desired size. The input {@link
```
A {@link PTransform} that batches inputs to a desired batch size.

echauchot replied:
A leftover again :) , thanks Thomas!

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,225 @@
+        BoundedWindow window) {
+      Instant firingInstant = window.maxTimestamp().plus(allowedLateness);
+      LOGGER.debug(
+          "*** SET TIMER *** to point in time %s for window %s",
```
Replace all Logger `%s` with `{}`

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,225 @@
+        @StateId(KEY_ID) ValueState<K> key,
+        ProcessContext c,
+        BoundedWindow window) {
+      Instant firingInstant = window.maxTimestamp().plus(allowedLateness);
```
`windowExpires`?

echauchot replied:
yes, it is exactly for window expiration. I coded it that way until `@OnWindowExpiration` is available.

tgroh replied:
Sorry for the lack of clarity.

Change the variable name? `windowExpires` is what this time is, and why you fire at that time is a consequence until `@OnWindowExpiration` is available.

echauchot replied:
ok, no problem. Thanks for the naming suggestion

tgroh commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java:
```diff
@@ -0,0 +1,232 @@
+    return data;
+  }
+
+  @Test
```
Add a test for outputting a batch of "the rest" at the time the window expires?

echauchot replied:
Isn't it what I do line 221 (batchSize is 5 and window contains 6 elements, so I test that it outputs 1 element)? Or do you mean testing that an element that comes between `window.maxTimestamp` and `window.maxTimestamp + allowedLateness` is correctly output in the batch?

tgroh replied:
Missed that. Thanks.

echauchot commented:
@tgroh, I addressed your comments except question above and rebased on master.
Besides, build seems broken:
```
Tests run: 16, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.477 sec <<< FAILURE! - in org.apache.beam.runners.direct.DirectRunnerTest
cancelShouldStopPipeline(org.apache.beam.runners.direct.DirectRunnerTest)  Time elapsed: 0.011 sec  <<< ERROR!
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@326c09f7 rejected from java.util.concurrent.ThreadPoolExecutor@36538b62[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 0]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
	at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:112)
	at org.apache.beam.runners.direct.TransformExecutorServices$ParallelEvaluationState.schedule(TransformExecutorServices.java:69)
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor.evaluateBundle(ExecutorServiceParallelExecutor.java:245)
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor.scheduleConsumption(ExecutorServiceParallelExecutor.java:208)
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor.access$1200(ExecutorServiceParallelExecutor.java:70)
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor$MonitorRunnable.addWorkIfNecessary(ExecutorServiceParallelExecutor.java:584)
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor$MonitorRunnable.run(ExecutorServiceParallelExecutor.java:482)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

```

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,226 @@
+ *  pipeline.run();
+ * }</pre>
+ */
+public class BatchingParDo<K, InputT>
```
This isn't really `BatchingParDo` anymore. `AssembleBatches`/`Batch`/`GroupIntoBatches`?

echauchot replied:
you're right, GroupIntoBatches pleases me

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java:
```diff
@@ -0,0 +1,226 @@
+    this.batchSize = batchSize;
+  }
+
+  public static <K, InputT> BatchingParDo<K, InputT> via(long batchSize) {
```
`ofSize(long batchSize)`

echauchot replied:
better name indeed, thanks

tgroh commented:
Thanks!

echauchot commented:
Yes of course, I left the detailed small commits for you to easily see each change requested.
Thanks Thomas!
